geopmctlpy/geopmctl/metric.py

```diff
@@ -0,0 +1,680 @@
+
+        trace_formatter (Formatter): Formatter for outputting the metric sample value in a trace
+        file. Passed to the Metric instance created from this signal as sample_only.
+        '''
+        self._descriptor = descriptor
+        self._behavior = behavior
+        if report_formatter is None:
+            self._report_formatter = Formatter(self.get_signature())
+        else:
+            self._report_formatter = report_formatter
+        if trace_formatter is None:
+            self._trace_formatter = Formatter(self.get_signature())
+        else:
+            self._trace_formatter = trace_formatter
+        self._value = None
+
+    def get_descriptor(self):
+        '''
+        Returns
+
+        A tuple of (name, domain, instance), which can be used in Agent.get_signals()
+        '''
+        return self._descriptor
+
+    def get_behavior(self):
+        '''
+        Getter method for Signal behavior.
+        '''
+        return self._behavior
+
+    def get_report_formatter(self):
+        '''
+        Getter method for Signal report formatter.
+        '''
+        return self._report_formatter
+
+    def get_trace_formatter(self):
+        '''
+        Getter method for Signal trace formatter.
+        '''
+        return self._trace_formatter
+
+    def value(self):
+        '''
+        Returns
+
+        Computed value for this signal according to the last update value. None, if there is not
+        enough information (for example, update() is not called yet).
+        '''
+        return self._value
+
+    def reset(self):
+        '''
+        Return to a state where update() has not been called yet.
+        '''
+        self._value = None
+
+    def update(self, value):
+        '''
+        Update the value of the signal from a sample.
+
+        Args
+
+        value (float): Value returned by pio.
+        '''
+        self._value = value
+
+    def get_signature(self):
+        '''
+        Returns
+
+        A string of name@domain-instance.
+        '''
+        return \
+            f'{self._descriptor[0]}@{topo.domain_name(self._descriptor[1])}-{self._descriptor[2]}'
+
+
+class MonotoneSignal(Signal):
+    '''
+    Implementation of Signal where the computed value is delta between the last sample and current
+    sample for monotone signals.
+    '''
+
+    def __init__(self, descriptor, report_formatter=None, trace_formatter=None):
+        self._last_value = None
+        super().__init__(descriptor, Behavior.MONOTONE, report_formatter=report_formatter,
+                         trace_formatter=trace_formatter)
+
+    def reset(self):
+        self._last_value = None
+        super().reset()
+
+    def update(self, value):
+        if self._last_value is None:
+            self._last_value = value
+        else:
+            self._value = value - self._last_value
+            self._last_value = value
+
+
+class Algorithm(ABC):
```

Comment:
Why do we need a separate object for the control algorithm for the Agent?

This really looks like the base class Agent to me.  Can you just use the runtime.py Agent class?